### PR TITLE
[https://nvbugs/6535765][fix] bump Wan layernorm fusion math to FP32

### PR DESCRIPTION
**Pull request by o-stoner**
## Summary

- Changed Wan 2.2 fused adaptive LayerNorm parameters from BF16 to FP32.
- Updated kernel loading, `AdaptiveLayerNormParams`, thop validation, and Wan call sites.
- Updated correctness tests for FP32 inputs and tighter tolerances.
- Improved fused-versus-golden LPIPS from `0.2393` to approximately `0.1878`.
- Maintained approximately `1.04×` speedup over eager execution.
- The `0.05` LPIPS threshold remains unmet. Visual generation tests remain expected to fail pending the remaining accuracy fix.

## Dev Engineer Review

- FP32 pointer types are consistent across the kernel API, thop validation, and Wan call sites.
- Kernel vector loads read LN affine and AdaLN modulation parameters as FP32.
- AdaLN scale transformation remains `1 + scale`. Shift values remain unchanged.
- No configuration or test-list files were modified.
- No end-to-end speed regression was reported.
- Follow-up work remains for the unmet LPIPS target.
- CI run `#68875` and downstream pipeline `#56261` passed after automatic retries.

## QA Engineer Review

- Modified `test_affine_fp32_weight_matches_eager()`.
- Removed `test_affine_fp32_weight_downcast_matches_eager()`.
- Updated modulation, validation, contiguity, and correctness test inputs to FP32.
- Coverage in `tests/integration/test_lists/` was not provided.
- Verdict: needs follow-up.

## Description

test_wan22_t2v_lpips_against_golden regressed after #15762 (Wan 2.2 LayerNorm + shift/scale + quant fusion), scoring LPIPS 0.2393 against a 0.05 threshold.

fusedAdaptiveLayerNormKernel read ln_weight/ln_bias/scale_msa/shift_msa as BF16, while the eager path it replaces computes AdaLN modulation in FP32, discarding precision in the modulation parameters before the kernel even runs. This PR changes the kernel AdaptiveLayerNormParams, the thop op, and the two WanBlock call sites to read these four tensors as FP32 directly, matching TRT-LLM's own layernormKernels.cu convention of FP32 accumulation for accuracy-sensitive math.

This lowers LPIPS (fused vs. golden) from 0.2393 -> ~0.188, but does not close the gap to the 0.05 threshold. Follow-up work is needed and tracked separately.

**LPIPS results:** {"fused_vs_golden" : 0.1878306269645691, "unfused_vs_golden" : 0.04006856679916382, "threshold" : 0.05}

Fused LN output mp4 (used to calculate LPIPS scores in CI):

https://github.com/user-attachments/assets/33768979-ece1-4c76-84e1-65868b60e58f

Unfused LN output mp4 (used to calculate LPIPS scores in CI):

https://github.com/user-attachments/assets/b25bb9f2-e6ca-4cfc-babd-0a5d83b74935

**Performance**

No E2E speed regression from this change. Re-ran the same NVFP4 720p × 81-frame, 40-step, torch.compile-on benchmark (3 runs) reported in #15762:

| | #15762 reported | This PR |
| --- | --- | --- |
| Fusion ON: Step time | 8.80 s | 8.65 s |
| Fusion ON: Total pipeline | 354.8 s | 348.0 s |
| Fusion OFF: Step time | 9.12 s | 9.03 s |
| Fusion OFF: Total pipeline | 367.5 s | 363.2 s |
| Speedup | 1.036× | 1.044× |

Fusion is still ~1.04× faster than eager, consistent with #15762.

## Test Coverage

- `tests/unittest/_torch/thop/parallel/test_fused_dit_layernorm_shift_scale.py:` updated kernel-level correctness tests to pass FP32 ln_weight/ln_bias/scale_msa/shift_msa matching the new conventions, tightened tolerances now that the downcast error is gone.
- `examples/visual_gen/test_visual_gen.py`::`test_wan22_t2v_lpips_against_golden` and the multi-GPU variants: the tests that caught this regression. Still expected to fail pending the follow-up fix for the remaining gap.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: cpp/tensorrt_llm/kernels/fusedAdaptiveLayerNormKernel.cu [hunk 1]
```diff
@@ -173,54 +173,68 @@ __launch_bounds__(BLOCK_SIZE, 8)
...
-                float2 bf = __bfloat1622float2(bVec2[i]);
-                wVals[chunk * 8 + i * 2 + 0] = wf.x;
-                wVals[chunk * 8 + i * 2 + 1] = wf.y;
-                bVals[chunk * 8 + i * 2 + 0] = bf.x;
-                bVals[chunk * 8 + i * 2 + 1] = bf.y;
-            }
+            float4 const w0 = *reinterpret_cast<float4 const*>(p.ln_weight + elemOff);
+            float4 const w1 = *reinterpret_cast<float4 const*>(p.ln_weight + elemOff + 4);
+            float4 const b0 = *reinterpret_cast<float4 const*>(p.ln_bias + elemOff);
+            float4 const b1 = *reinterpret_cast<float4 const*>(p.ln_bias + elemOff + 4);
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply cpp/tensorrt_llm/kernels/fusedAdaptiveLayerNormKernel.cu:186]: The same pattern (is_contiguous() + dtype check, no explicit alignment check) is used across all sibling ops. This is the established convention in this codebase and the error path is unreachable in the current codebase (the kernel requires D % 16 == 0 (elements), and the op hard-enforces D == 5120); will not fix

File: tensorrt_llm/_torch/visual_gen/models/wan/transformer_wan.py [hunk 2]
```diff
@@ -573,8 +573,8 @@ def forward(
             _x_2d = x.reshape(-1, x.shape[-1])
             norm_x = apply_fused_layernorm_affine_quant(
                 _x_2d,
-                self.norm2.weight.to(x.dtype),
-                self.norm2.bias.to(x.dtype),
+                self.norm2.weight,
+                self.norm2.bias,
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tensorrt_llm/_torch/visual_gen/models/wan/transformer_wan.py]: Not applicable; visual_gen isn't in the mypy-checked file set (pyproject.toml [tool.mypy] files, which static typing is opt-in per CODING_GUIDELINES.md), and the flagged lines are call arguments in WanBlock.forward, a pre-existing, already-unannotated signature untouched by this PR

File: tests/unittest/_torch/visual_gen/kernels/parallel/test_fused_dit_layernorm_shift_scale.py [hunk 3]
```diff
@@ -544,11 +523,14 @@ def test_adaln_production_modulator_extraction():
...
-    ref = _ref_layernorm_adaln(x, scale_msa, shift_msa, S, EPS).to(torch.bfloat16)
+    # utils_wan.py casts to float32 before the kernel; mimic that here.
+    out = torch.ops.trtllm.fused_adaptive_layernorm(
+        x, None, None, scale_msa.float(), shift_msa.float(), S, EPS
+    )
+    ref = _ref_layernorm_adaln(x, scale_msa, shift_msa, S, EPS)
 
     assert out.shape == (M, D)
-    torch.testing.assert_close(out.float(), ref.float(), rtol=2e-2, atol=2e-2)
+    torch.testing.assert_close(out.float(), ref.float(), rtol=1e-2, atol=1e-2)
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tests/unittest/_torch/visual_gen/kernels/parallel/test_fused_dit_layernorm_shift_scale.py:533]: Not adding this; it would be inconsistent with the established convention in this test-file family. `test_adaln_production_modulator_extraction` already mimics the wrapper's exact production transform pipeline (chunk -> squeeze -> reshape -> .contiguous() -> float cast) inline and asserts against it. The sibling test file for the analogous LTX2 wrappers (`test_fused_dit_gate_resid_norm_shift_scale.py`, covering `utils_ltx2.py's apply_fused_rmsnorm_shift_scale` etc.) follows the same pattern; those wrappers are also never imported/called directly, only referenced in comments. The wrapper functions themselves are thin (dtype cast + .contiguous() + Fp4QuantizedTensor packaging), and the full `WanBlock.forward` -> `apply_fused_layernorm_*` path is exercised end-to-end by `test_wan_transformer.py`::`test_cosine_similarity` at the 5120 hidden-size (14B) config.

**Review by karljang — APPROVED** [review]
LGTM overall. The FP32 change looks consistent and improves the existing LPIPS regression without a performance regression.

A few non-blocking follow-ups: consider validating 16-byte alignment for the `float4` loads, strengthen the FP32 test with values that clearly distinguish it from BF16 downcasting, and link the follow-up tracking the remaining LPIPS gap.

File: tests/unittest/_torch/visual_gen/kernels/parallel/test_fused_dit_layernorm_shift_scale.py [hunk 4]
```diff
@@ -149,36 +146,29 @@ def test_adaln_batch_modulation_correctness():
...
-# reads them as bf16. Verify the downcast matches the float32-weight reference.
+# Affine FP32 weight/bias: kernel reads them in fp32; verify the output matches
+# the FP32 reference closely (not limited by a bf16 downcast of weights).
 # ---------------------------------------------------------------------------
 
 
-def test_affine_fp32_weight_downcast_matches_eager():
-    """fp32 weight/bias downcast to bf16 matches the float32-weight reference."""
+def test_affine_fp32_weight_matches_eager():
+    """fp32 weight/bias passed directly to kernel match the fp32 reference."""
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tests/unittest/_torch/visual_gen/kernels/parallel/test_fused_dit_layernorm_shift_scale.py:171]: Every other test function in this file (and its sibling files across `tests/unittest/_torch/thop/parallel/`) is unannotated, so adding `-> None` to just this one function would be inconsistent with the rest of the file. Will not fix.

File: tests/unittest/_torch/visual_gen/kernels/parallel/test_fused_dit_layernorm_shift_scale.py [hunk 5]
```diff
@@ -75,15 +75,13 @@ def _make_inputs(M, B, has_ln_affine, has_modulation, seed=42):
...
-        ln_bias = (torch.randn(D, device=device) * 0.1).to(torch.bfloat16)
+        ln_weight = torch.ones(D, device=device) + torch.randn(D, device=device) * 0.1
+        ln_bias = torch.randn(D, device=device) * 0.1
         ref = _ref_layernorm_affine(x, ln_weight, ln_bias, EPS)
     elif has_modulation:
         seq_len_per_batch = M // B
-        scale_msa = (torch.randn(B, D, device=device) * 0.2).to(torch.bfloat16)
-        shift_msa = (torch.randn(B, D, device=device) * 0.2).to(torch.bfloat16)
+        scale_msa = torch.randn(B, D, device=device) * 0.2
+        shift_msa = torch.randn(B, D, device=device) * 0.2
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tests/unittest/_torch/visual_gen/kernels/parallel/test_fused_dit_layernorm_shift_scale.py:84]: Fixed; the test now asserts that the kernel output is closer to fp32 reference

File: tests/unittest/_torch/visual_gen/kernels/parallel/test_fused_dit_layernorm_shift_scale.py [hunk 6]
```diff
@@ -149,36 +146,29 @@ def test_adaln_batch_modulation_correctness():
...
-    # Cross-batch: rows from different batch elements should be measurably different.
     out_b0 = out[:S].float().mean()
     out_b1 = out[S : 2 * S].float().mean()
     assert abs(out_b0.item() - out_b1.item()) > 0.1, (
         "Batch elements have indistinguishable outputs — modulation broadcast may be broken"
     )
 
-    # Still matches reference row-by-row.
-    torch.testing.assert_close(out.float(), ref.float(), rtol=2e-2, atol=2e-2)
+    torch.testing.assert_close(out.float(), ref.float(), rtol=1e-2, atol=1e-2)
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tests/unittest/_torch/visual_gen/kernels/parallel/test_fused_dit_layernorm_shift_scale.py:161]: Unreachable in practice, so I'm skipping - `scale_msa`/`shift_msa` go through `apply_fused_layernorm_adaln_quant` in `utils_wan.py`, which unconditionally does `.to(dtype=torch.float32)` on both tensors immediately before calling the op, regardless of the incoming dtype. And on the C++ side, `fusedAdaptiveLayerNormOp.cpp`::`validateInputs()` has a hard `CHECK_TYPE(scale_msa, torch::kFloat32)` / `CHECK_TYPE(shift_msa, torch::kFloat32)` assert; if a bf16 tensor ever reached the op, it would throw immediately rather than silently downcast. I applied the same reasoning to revert an earlier commit that added an equivalent check for `ln_weight`/`ln_bias` (same double-guard there too).

File: tests/unittest/_torch/visual_gen/kernels/parallel/test_fused_dit_layernorm_shift_scale.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (o-stoner) [reply tests/unittest/_torch/visual_gen/kernels/parallel/test_fused_dit_layernorm_shift_scale.py:84]: Reverted this based on the conversation in https://github.com/NVIDIA/TensorRT-LLM/pull/17987#discussion_r3882449210 .

**Review by kris1025 — APPROVED** [review]
